File: stack/mini-stack.py
from collections import deque

class MinStack:
    
    def __init__(self):
        self.stack = deque()
        self.mini_stack = deque()

    def push(self, x: int) -> None:
        self.stack.append(x)
        if self.getMin() is not None:
            m = min(x, self.getMin())
            if x == m:
                self.mini_stack.append(m)
        else:
            self.mini_stack.append(x)

    def pop(self) -> None:
        p = self.stack.pop()
        if p == self.getMin():
            self.mini_stack.pop()

    def top(self) -> int:
        return self.stack[-1] if len(self.stack) > 0 else None

    def getMin(self) -> int:
        return self.mini_stack[-1] if len(self.mini_stack) > 0 else None


obj = MinStack()
# obj.push(-2)
# obj.push(0)
# obj.push(-3)
# print(obj.getMin())
# obj.pop()
# print(obj.top())
# print(obj.getMin())

obj.push(2)
obj.push(0)
obj.push(3)
obj.push(0)
print(obj.getMin())
obj.pop()
print(obj.getMin())
obj.pop()
print(obj.getMin())
obj.pop()
print(obj.getMin())

File: stack/mini-stack.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.stack = [],
    this.mini_stack = []
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    // console.log('x', x)
    this.stack.push(x)
    let p = x;
    let flag = true
    if(this.getMin() != null && this.getMin() != undefined){
        p = Math.min(this.getMin(), x);
        if(x > this.getMin()){
            flag = false
        }
    }
    if(flag){
        this.mini_stack.push(p)
    }
    // console.log('p', p)
    // console.log(this.mini_stack)
    // console.log(this.stack)
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    let p = this.stack.pop()
    // console.log('p', p)
    // console.log('mini', this.getMin())
    if(p == this.getMin()){
        this.mini_stack.pop()
        // console.log(this.mini_stack)
    }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    let s = this.stack.slice()
    return s.reverse()[0]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    let m = this.mini_stack.slice()
    return m.reverse()[0]
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

var obj = new MinStack()
// obj.push(-2)
// obj.push(0)
// obj.push(-3)
// console.log(obj.getMin())
// obj.pop()
// console.log(obj.top())
// console.log(obj.getMin())

// obj.push(0)
// obj.push(1)
// obj.push(0)
// console.log(obj.getMin())
// obj.pop()
// console.log(obj.getMin())

obj.push(2)
obj.push(0)
obj.push(3)
obj.push(0)
console.log(obj.getMin())
obj.pop()
console.log(obj.getMin())
obj.pop()
console.log(obj.getMin())
obj.pop()
console.log(obj.getMin())